allennlp/modules/seq2seq_encoders/bidirectional_transformer_encoder.py

```diff
@@ -0,0 +1,265 @@
+    if mask is not None:
+        scores = scores.masked_fill(mask == 0, -1e9)
+    p_attn = F.softmax(scores, dim=-1)
+    if dropout is not None:
+        p_attn = dropout(p_attn)
+    return torch.matmul(p_attn, value), p_attn
+
+def subsequent_mask(size: int, device: str = 'cpu') -> torch.Tensor:
+    """Mask out subsequent positions."""
+    attn_shape = (1, size, size)
+    mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
+    mask = (torch.from_numpy(mask) == 0)
+    return mask.to(device)
+
+
+class PositionalEncoding(torch.nn.Module):
+    "Implement the Positional Encoding function."
+    def __init__(self, input_dim: int, max_len: int = 5000) -> None:
+        super().__init__()
+
+        # Compute the positional encodings once in log space.
+        positional_encoding = torch.zeros(max_len, input_dim, requires_grad=False)
+        position = torch.arange(0, max_len).unsqueeze(1).float()
+        div_term = torch.exp(torch.arange(0, input_dim, 2).float() * -(math.log(10000.0) / input_dim))
+        positional_encoding[:, 0::2] = torch.sin(position * div_term)
+        positional_encoding[:, 1::2] = torch.cos(position * div_term)
+        positional_encoding = positional_encoding.unsqueeze(0)
+        self.register_buffer('positional_encoding', positional_encoding)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=arguments-differ
+        return x + self.positional_encoding[:, :x.size(1)]
+
+
+class PositionwiseFeedForward(torch.nn.Module):
+    "Implements FFN equation."
+    def __init__(self, input_dim: int, ff_dim: int, dropout: float = 0.1) -> None:
+        super().__init__()
+        self.w_1 = torch.nn.Linear(input_dim, ff_dim)
+        self.w_2 = torch.nn.Linear(ff_dim, input_dim)
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=arguments-differ
+        return self.w_2(self.dropout(F.relu(self.w_1(x))))
+
+
+class TransformerEncoder(torch.nn.Module):
+    "Core encoder is a stack of N layers"
+    def __init__(self, layer: torch.nn.Module, num_layers: int, return_all_layers: bool = False) -> None:
+        super().__init__()
+        self.layers = util.clone(layer, num_layers)
+        self.norm = LayerNorm(layer.size)
+        self.return_all_layers = return_all_layers
+
+    def forward(self, x, mask):
+        "Pass the input (and mask) through each layer in turn."
+        all_layers = []
+        for layer in self.layers:
+            x = layer(x, mask)
+            if self.return_all_layers:
+                all_layers.append(x)
+
+        if self.return_all_layers:
+            all_layers[-1] = self.norm(all_layers[-1])
+            return all_layers
+        return self.norm(x)
+
+
+class SublayerConnection(torch.nn.Module):
+    """
+    A residual connection followed by a layer norm.
+    Note for code simplicity the norm is first as opposed to last.
```

Comment:
Huh. This is different than the original transformer described in "Attention Is All You Need". They use `self.norm(x + self.dropout(sublayer(x)))`. Not sure why that's more complicated. Regardless, since this is how Calypso/Annotated Transformer do it, looks good...

Reply:
Ah, it's simply that they apply the norm later. Not that it's left out.